dspace-api/src/main/java/org/dspace/uclouvain/citations/UCLouvainCitationsServiceImpl.java

```diff
@@ -0,0 +1,154 @@
+import org.dspace.content.service.ItemService;
+import org.dspace.core.Context;
+import org.dspace.utils.DSpace;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * Main service to generate text citations for an item.
+ * We use this instead of the classic DSpace system which is using processes and returns files to the user.
+ * 
+ * @author Michaël Pourbaix (michael.pourbaix@uclouvain.be)
+ */
+public class UCLouvainCitationsServiceImpl implements UCLouvainCitationsService {
+    private Logger logger = LogManager.getLogger(UCLouvainCitationsServiceImpl.class);
+    private StreamDisseminationCrosswalkMapper crosswalkMapper;
+    private static final String CSL_ERROR_REGEX = "\\[CSL STYLE ERROR: .*\\]";
+
+    @Autowired
+    ItemExportFormatService itemExportFormatService;
+
+    @Autowired
+    ItemService itemService;
+
+    @PostConstruct
+    void init() {
+        this.crosswalkMapper =
+            new DSpace().getSingletonService(StreamDisseminationCrosswalkMapper.class);
+    }
+
+    public boolean isValidFormat(Context context, String citationFormat) {
+        return this.crosswalkMapper.getTypes().contains(citationFormat);
+    }
+
+    /**
+     * Generate a citation for a given format and item.
+     * 
+     * @param context The current DSpace context.
+     * @param item The item to generated a citation for.
+     * @param citationFormat The desired format of the citation.
+     * 
+     */
+    public ItemCitation getCitationForItem(Context context, Item item, String citationFormat)
+        throws UnknownCitationFormatException {
+        // Retrieve the crosswalk for the given format.
+        StreamDisseminationCrosswalk citationCrosswalk = this.crosswalkMapper.getByType(citationFormat);
+        if (citationCrosswalk == null) {
+            throw new UnknownCitationFormatException(
+                "The citation format '" + citationFormat + "' is not configured in the system."
+            );
+        }
+
+        if (!citationCrosswalk.canDisseminate(context, item)) {
+            return null;
+        }
+        try {
+            // Write the result in an output stream and then extract the content from it.
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+            citationCrosswalk.disseminate(context, item, output);
+            String citationResult = output.toString(StandardCharsets.UTF_8);
+            System.out.println("Processed citation, got: " + citationResult);
+            // The processor may return an error in certain cases.
+            if (containsError(citationResult)) {
+                // If we detect an error: log it and replace the citation result with null.
+                logger.warn(
+                    "Could not generate citation of format: [" + citationFormat + "] for item: [" + item.getID() + "]."
+                    + " Error is: '" + citationResult.trim() + "'."
+                );
+                citationResult = null;
```

Comment:
could be interest to return a debug line such `String.format(Citation({}) generation error for {}", citationFormat, item.getID())` ?